In [1]:
import pandas as pd
import numpy as np
import statsmodels.formula.api as smf

# 1. Explore the Data

## 1.1

In [2]:
data = pd.read_csv('./wdbc.csv.bz2')

In [3]:
data.head()

id diagnosis  radius.mean  texture.mean  perimeter.mean  area.mean  \
0    842302         M        17.99         10.38          122.80     1001.0   
1    842517         M        20.57         17.77          132.90     1326.0   
2  84300903         M        19.69         21.25          130.00     1203.0   
3  84348301         M        11.42         20.38           77.58      386.1   
4  84358402         M        20.29         14.34          135.10     1297.0   

   smoothness.mean  compactness.mean  concavity.mean  concpoints.mean  \
0          0.11840           0.27760          0.3001          0.14710   
1          0.08474           0.07864          0.0869          0.07017   
2          0.10960           0.15990          0.1974          0.12790   
3          0.14250           0.28390          0.2414          0.10520   
4          0.10030           0.13280          0.1980          0.10430   

       ...        radius.worst  texture.worst  perimeter.worst  area.worst  \
0      ...               25.38          17.33           184.60      2019.0   
1      ...               24.99          23.41           158.80      1956.0   
2      ...               23.57          25.53           152.50      1709.0   
3      ...               14.91          26.50            98.87       567.7   
4      ...               22.54          16.67           152.20      1575.0   

   smoothness.worst  compactness.worst  concavity.worst  concpoints.worst  \
0            0.1622             0.6656           0.7119            0.2654   
1            0.1238             0.1866           0.2416            0.1860   
2            0.1444             0.4245           0.4504            0.2430   
3            0.2098             0.8663           0.6869            0.2575   
4            0.1374             0.2050           0.4000            0.1625   

   symmetry.worst  fracdim.worst  
0          0.4601        0.11890  
1          0.2750        0.08902  
2          0.3613        0.08758  
3          0.6638        0.17300  
4          0.2364        0.07678  

[5 rows x 32 columns]

In [4]:
data.diagnosis.value_counts()

B    357
M    212
Name: diagnosis, dtype: int64

In [5]:
data.diagnosis = (data.diagnosis == 'M').astype(int)

In [6]:
data.head()

id  diagnosis  radius.mean  texture.mean  perimeter.mean  area.mean  \
0    842302          1        17.99         10.38          122.80     1001.0   
1    842517          1        20.57         17.77          132.90     1326.0   
2  84300903          1        19.69         21.25          130.00     1203.0   
3  84348301          1        11.42         20.38           77.58      386.1   
4  84358402          1        20.29         14.34          135.10     1297.0   

   smoothness.mean  compactness.mean  concavity.mean  concpoints.mean  \
0          0.11840           0.27760          0.3001          0.14710   
1          0.08474           0.07864          0.0869          0.07017   
2          0.10960           0.15990          0.1974          0.12790   
3          0.14250           0.28390          0.2414          0.10520   
4          0.10030           0.13280          0.1980          0.10430   

       ...        radius.worst  texture.worst  perimeter.worst  area.worst  \
0      ...               25.38          17.33           184.60      2019.0   
1      ...               24.99          23.41           158.80      1956.0   
2      ...               23.57          25.53           152.50      1709.0   
3      ...               14.91          26.50            98.87       567.7   
4      ...               22.54          16.67           152.20      1575.0   

   smoothness.worst  compactness.worst  concavity.worst  concpoints.worst  \
0            0.1622             0.6656           0.7119            0.2654   
1            0.1238             0.1866           0.2416            0.1860   
2            0.1444             0.4245           0.4504            0.2430   
3            0.2098             0.8663           0.6869            0.2575   
4            0.1374             0.2050           0.4000            0.1625   

   symmetry.worst  fracdim.worst  
0          0.4601        0.11890  
1          0.2750        0.08902  
2          0.3613        0.08758  
3          0.6638        0.17300  
4          0.2364        0.07678  

[5 rows x 32 columns]

## 1.2

In [7]:
summary = pd.DataFrame({'variable' : [], 'correlation' : [], 'mean' : [], 'min' : [], 'max' : [], 'std' : [], 'missings' : []})

In [8]:
summary

Empty DataFrame
Columns: [variable, correlation, mean, min, max, std, missings]
Index: []

In [9]:
summary.variable = data.columns

In [10]:
summary

variable  correlation  mean  min  max  std  missings
0                  id          NaN   NaN  NaN  NaN  NaN       NaN
1           diagnosis          NaN   NaN  NaN  NaN  NaN       NaN
2         radius.mean          NaN   NaN  NaN  NaN  NaN       NaN
3        texture.mean          NaN   NaN  NaN  NaN  NaN       NaN
4      perimeter.mean          NaN   NaN  NaN  NaN  NaN       NaN
5           area.mean          NaN   NaN  NaN  NaN  NaN       NaN
6     smoothness.mean          NaN   NaN  NaN  NaN  NaN       NaN
7    compactness.mean          NaN   NaN  NaN  NaN  NaN       NaN
8      concavity.mean          NaN   NaN  NaN  NaN  NaN       NaN
9     concpoints.mean          NaN   NaN  NaN  NaN  NaN       NaN
10      symmetry.mean          NaN   NaN  NaN  NaN  NaN       NaN
11       fracdim.mean          NaN   NaN  NaN  NaN  NaN       NaN
12          radius.se          NaN   NaN  NaN  NaN  NaN       NaN
13         texture.se          NaN   NaN  NaN  NaN  NaN       NaN
14       perimeter.se          NaN   NaN  NaN  NaN  NaN       NaN
15            area.se          NaN   NaN  NaN  NaN  NaN       NaN
16      smoothness.se          NaN   NaN  NaN  NaN  NaN       NaN
17     compactness.se          NaN   NaN  NaN  NaN  NaN       NaN
18       concavity.se          NaN   NaN  NaN  NaN  NaN       NaN
19      concpoints.se          NaN   NaN  NaN  NaN  NaN       NaN
20        symmetry.se          NaN   NaN  NaN  NaN  NaN       NaN
21         fracdim.se          NaN   NaN  NaN  NaN  NaN       NaN
22       radius.worst          NaN   NaN  NaN  NaN  NaN       NaN
23      texture.worst          NaN   NaN  NaN  NaN  NaN       NaN
24    perimeter.worst          NaN   NaN  NaN  NaN  NaN       NaN
25         area.worst          NaN   NaN  NaN  NaN  NaN       NaN
26   smoothness.worst          NaN   NaN  NaN  NaN  NaN       NaN
27  compactness.worst          NaN   NaN  NaN  NaN  NaN       NaN
28    concavity.worst          NaN   NaN  NaN  NaN  NaN       NaN
29   concpoints.worst          NaN   NaN  NaN  NaN  NaN       NaN
30     symmetry.worst          NaN   NaN  NaN  NaN  NaN       NaN
31      fracdim.worst          NaN   NaN  NaN  NaN  NaN       NaN

In [11]:
for i in summary.variable:
    index = summary.index[summary['variable'] == i]
    summary.loc[index] = [i, data['diagnosis'].corr(data[i], method='pearson'), np.mean(data[i]), np.min(data[i]), np.max(data[i]), np.std(data[i]), data[i].isnull().sum()]



In [12]:
summary

variable  correlation          mean          min           max  \
0                  id     0.039769  3.037183e+07  8670.000000  9.113205e+08   
1           diagnosis     1.000000  3.725835e-01     0.000000  1.000000e+00   
2         radius.mean     0.730029  1.412729e+01     6.981000  2.811000e+01   
3        texture.mean     0.415185  1.928965e+01     9.710000  3.928000e+01   
4      perimeter.mean     0.742636  9.196903e+01    43.790000  1.885000e+02   
5           area.mean     0.708984  6.548891e+02   143.500000  2.501000e+03   
6     smoothness.mean     0.358560  9.636028e-02     0.052630  1.634000e-01   
7    compactness.mean     0.596534  1.043410e-01     0.019380  3.454000e-01   
8      concavity.mean     0.696360  8.879932e-02     0.000000  4.268000e-01   
9     concpoints.mean     0.776614  4.891915e-02     0.000000  2.012000e-01   
10      symmetry.mean     0.330499  1.811619e-01     0.106000  3.040000e-01   
11       fracdim.mean    -0.012838  6.279761e-02     0.049960  9.744000e-02   
12          radius.se     0.567134  4.051721e-01     0.111500  2.873000e+00   
13         texture.se    -0.008303  1.216853e+00     0.360200  4.885000e+00   
14       perimeter.se     0.556141  2.866059e+00     0.757000  2.198000e+01   
15            area.se     0.548236  4.033708e+01     6.802000  5.422000e+02   
16      smoothness.se    -0.067016  7.040979e-03     0.001713  3.113000e-02   
17     compactness.se     0.292999  2.547814e-02     0.002252  1.354000e-01   
18       concavity.se     0.253730  3.189372e-02     0.000000  3.960000e-01   
19      concpoints.se     0.408042  1.179614e-02     0.000000  5.279000e-02   
20        symmetry.se    -0.006522  2.054230e-02     0.007882  7.895000e-02   
21         fracdim.se     0.077972  3.794904e-03     0.000895  2.984000e-02   
22       radius.worst     0.776454  1.626919e+01     7.930000  3.604000e+01   
23      texture.worst     0.456903  2.567722e+01    12.020000  4.954000e+01   
24    perimeter.worst     0.782914  1.072612e+02    50.410000  2.512000e+02   
25         area.worst     0.733825  8.805831e+02   185.200000  4.254000e+03   
26   smoothness.worst     0.421465  1.323686e-01     0.071170  2.226000e-01   
27  compactness.worst     0.590998  2.542650e-01     0.027290  1.058000e+00   
28    concavity.worst     0.659610  2.721885e-01     0.000000  1.252000e+00   
29   concpoints.worst     0.793566  1.146062e-01     0.000000  2.910000e-01   
30     symmetry.worst     0.416294  2.900756e-01     0.156500  6.638000e-01   
31      fracdim.worst     0.323872  8.394582e-02     0.055040  2.075000e-01   

             std  missings  
0   1.249107e+08       0.0  
1   4.834925e-01       0.0  
2   3.520951e+00       0.0  
3   4.297255e+00       0.0  
4   2.427762e+01       0.0  
5   3.516048e+02       0.0  
6   1.405176e-02       0.0  
7   5.276633e-02       0.0  
8   7.964973e-02       0.0  
9   3.876873e-02       0.0  
10  2.739018e-02       0.0  
11  7.054156e-03       0.0  
12  2.770689e-01       0.0  
13  5.511634e-01       0.0  
14  2.020077e+00       0.0  
15  4.545101e+01       0.0  
16  2.999878e-03       0.0  
17  1.789244e-02       0.0  
18  3.015952e-02       0.0  
19  6.164861e-03       0.0  
20  8.259104e-03       0.0  
21  2.643745e-03       0.0  
22  4.828993e+00       0.0  
23  6.140854e+00       0.0  
24  3.357300e+01       0.0  
25  5.688565e+02       0.0  
26  2.281236e-02       0.0  
27  1.571982e-01       0.0  
28  2.084409e-01       0.0  
29  6.567455e-02       0.0  
30  6.181308e-02       0.0  
31  1.804539e-02       0.0

## 2. Which Model is Best?

In [13]:
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split

## 2.1

In [14]:
X = data[['concpoints.mean', 'fracdim.mean']].values
y = data.diagnosis

In [15]:
X.shape

(569, 2)

In [16]:
Xtrain, Xvalid, ytrain, yvalid = train_test_split(X, y, test_size = 0.2)

## 2.2

In [17]:
m = LogisticRegression(solver='lbfgs', C=1e9).fit(Xtrain, ytrain)

## 2.3

In [18]:
Xtrain.shape

(455, 2)

In [19]:
yhat = m.predict(Xtrain)

In [20]:
rmse = np.sqrt(np.mean((ytrain - yhat)**2))
rmse

0.28516430582855445

In [21]:
y.shape

(569,)

In [22]:
yhat.shape

(455,)

In [23]:
pd.crosstab(ytrain, yhat)

col_0        0    1
diagnosis          
0          275   16
1           21  143

## 2.4

In [24]:
a = (150 + 273) / 455
a

0.9296703296703297

the accuracy is ~93%

In [25]:
p = 150 / (150 + 14)

In [26]:
r = 150 / (150 + 18)

In [27]:
f = 2 / (1 / p + 1 / r)
f

0.9036144578313254

the f score is 0.903

## 2.5

In [28]:
yhat = m.predict(Xvalid)

In [29]:
rmse = np.sqrt(np.mean((yvalid - yhat)**2))
rmse

0.28097574347450816

In [30]:
yhat.shape

(114,)

In [31]:
pd.crosstab(yvalid, yhat)

col_0       0   1
diagnosis        
0          62   4
1           5  43

## 2.6

In [32]:
a = (35 + 65) / 114
a

0.8771929824561403

the accuracy is 87.7%

In [33]:
r = 35 / (35 + 9)

In [34]:
p = 35 / (35 + 5)

In [35]:
f = 2 / (1 / p + 1 / r)
f

0.8333333333333334

the f score is 0.833

## 2.7

The model performs better on training data, evidenced by higher accuracy and a lower RMSE. However, there is not a drastic difference between the RMSE's, indicating that the model it not overfitting.

## 2.8

### All variables

In [36]:
X0 = data.copy().drop(['diagnosis', 'id'], axis=1)

In [37]:
Xtrain, Xvalid, ytrain, yvalid = train_test_split(X0, y, test_size = 0.2)

In [38]:
m = LogisticRegression(solver='lbfgs', C=1e9).fit(Xtrain, ytrain)

/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:758: ConvergenceWarning: lbfgs failed to converge. Increase the number of iterations.
  "of iterations.", ConvergenceWarning)


In [39]:
m.coef_

array([[-1.90278825e+00, -6.35119250e-01,  1.50824754e-01,
        -1.91266335e-03,  7.66041628e-02,  3.48569635e-01,
         4.84612756e-01,  2.08980307e-01,  1.11974085e-01,
         2.22327983e-02, -8.34278985e-02, -9.91243263e-01,
        -4.75438838e-01,  1.56888191e-01,  7.25293177e-03,
         6.30303767e-02,  9.30468953e-02,  2.69229236e-02,
         2.50263254e-02,  5.97562400e-03, -1.99608246e+00,
         6.28547938e-01,  1.85299707e-01,  2.33506534e-02,
         1.37904354e-01,  1.05578295e+00,  1.29959951e+00,
         3.98142463e-01,  3.40203605e-01,  1.04386003e-01]])

In [40]:
yhat = m.predict(Xtrain)

In [41]:
np.unique(yhat)

array([0, 1])

In [42]:
rmse = np.sqrt(np.mean((ytrain - yhat)**2))
rmse

0.22966770070528583

In [43]:
yhat.shape

(455,)

In [44]:
pd.crosstab(ytrain, yhat)

col_0        0    1
diagnosis          
0          273   10
1           14  158

In [45]:
a = (159 + 278) / 455
a

0.9604395604395605

the accuracy of this model is 96%

In [46]:
r = 159 / (159 + 13)

In [47]:
p = 159 / (159 + 5)

In [48]:
F = 2 / (1 / p + 1 / r)
F

0.9464285714285714

The f score of this model is 0.946

In [49]:
yhat = m.predict(Xvalid)

In [50]:
rmse = np.sqrt(np.mean((yvalid - yhat)**2))
rmse

0.20942695414584775

In [51]:
pd.crosstab(yvalid, yhat)

col_0       0   1
diagnosis        
0          73   1
1           4  36

In [52]:
yhat.shape

(114,)

In [66]:
a = (36 + 73) / 114
a

0.956140350877193

the accuracy of this model is 95.6%

In [54]:
r = 38 / (38 + 2)

In [55]:
p = 38 / (38 + 2)

In [56]:
f = 2 / (1 / p + 1 / r)
f

0.9500000000000001

the f score of this model is 0.95

### Radius, Texture, Area, Smoothness

In [57]:
X1 = data[['radius.mean', 'texture.mean', 'area.mean', 'smoothness.mean']]

In [58]:
Xtrain, Xvalid, ytrain, yvalid = train_test_split(X1, y, test_size = 0.2)

In [59]:
m = LogisticRegression(solver='lbfgs', C=1e9).fit(Xtrain, ytrain)

/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:758: ConvergenceWarning: lbfgs failed to converge. Increase the number of iterations.
  "of iterations.", ConvergenceWarning)


In [60]:
yhat = m.predict(Xtrain)

In [61]:
rmse = np.sqrt(np.mean((ytrain - yhat)**2))
rmse

0.2693092507276208

In [62]:
yhat.shape

(455,)

In [63]:
pd.crosstab(ytrain, yhat)

col_0        0    1
diagnosis          
0          269   15
1           18  153

In [65]:
a = (153 + 269) / 455
a

0.9274725274725275

The accuracy of this model is 92.7%

In [67]:
r = 153 / (153 + 18)

In [68]:
p = 153 / (153 + 15)

In [69]:
F = 2 / (1 / p + 1 / r)
F

0.9026548672566371

The F score of this model is 90.3%

In [70]:
yhat = m.predict(Xvalid)
rmse = np.sqrt(np.mean((yvalid - yhat)**2))
rmse

0.31063037209869776

In [72]:
pd.crosstab(yvalid, yhat)

col_0       0   1
diagnosis        
0          69   4
1           7  34

In [73]:
a = (34 + 69) / 114
a

0.9035087719298246

The accuracy of this model is 90.4%

In [74]:
r = 34 / (34 + 7)

In [75]:
p = 34 / (34 + 4)

In [76]:
F = 2 / (1 / p + 1 / r)
F

0.860759493670886

The F score of this model is 0.86

## 2.9

It seems that the model using all the variables performs the best, in terms of RMSE and accuracy.